sw/host/opentitanlib/src/transport/ti50emulator/i2c.rs

```diff
@@ -2,14 +2,268 @@
+            inner: Rc::clone(&ti50.inner),
+            socket: RefCell::default(),
+            path: soc_path,
+            last_id: Cell::new(EMULATOR_INVALID_ID),
+        })
+    }
+
+    /// Function re-connect socket to `SubProcess` when detect
+    /// that process was restarted.
+    pub fn reconnect(&self) -> Result<()> {
+        let mut socket = self.socket.borrow_mut();
+        let id = self.inner.borrow_mut().process.get_id();
+        if self.last_id.get() != id {
+            *socket = Some(UnixStream::connect(&self.path)?);
+            self.last_id.set(id);
+        }
+        Ok(())
+    }
+
+    /// Function check if I2C transaction should be performed on DUT.
+    pub fn check_state(&self) -> Result<()> {
+        let process = &mut self.inner.borrow_mut().process;
+        process.update_status()?;
+        match process.get_state() {
+            EmuState::On => Ok(()),
+            state => Err(I2cError::Generic(format!(
+                "Operation not supported in Emulator state: {}",
+                state
+            ))
+            .into()),
+        }
+    }
+
+    /// Function try to receive data from unix socket.
+    pub fn rx(fd: &mut UnixStream, data: &mut [u8], deadline: Instant) -> Result<usize> {
+        let mut rx_count: usize = 0;
+        while rx_count < data.len() {
+            if Instant::now() > deadline {
+                bail!(I2cError::Timeout);
+            }
+            match fd.read(&mut data[rx_count..]) {
+                Ok(0) => {
+                    bail!(I2cError::Generic("Connection closed".to_string()));
+                }
+                Ok(n) => {
+                    rx_count += n;
+                }
+                Err(err) => {
+                    if err.kind() != ErrorKind::WouldBlock {
+                        return Err(err.into());
+                    }
+                }
+            }
+        }
+        Ok(rx_count)
+    }
+
+    /// Function send contents of slice to unix socket.
+    pub fn tx(fd: &mut UnixStream, data: &[u8], deadline: Instant) -> Result<()> {
+        let mut tx_count: usize = 0;
+        while tx_count < data.len() {
+            if Instant::now() > deadline {
+                bail!(I2cError::Timeout);
+            }
+            match fd.write(&data[tx_count..data.len()]) {
+                Ok(0) => {
+                    bail!(I2cError::Generic("Connection closed".to_string()));
+                }
+                Ok(n) => {
+                    tx_count += n;
+                }
+                Err(err) => {
+                    if err.kind() != ErrorKind::WouldBlock {
+                        return Err(err.into());
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Function perform write transaction on emulated bus.
+    pub fn write(&self, addr: u8, data: &[u8]) -> Result<()> {
+        if let Some(ref mut fd) = *self.socket.borrow_mut() {
+            fd.set_nonblocking(true)?;
```

Comment:
What is the point of making the socket non-blocking in write() but blocking in read()? when both use the same underlying tx() and rx() which handles timeouts?  I would have assumed that nonblocking works for both, and could be set in reconnect().

Reply:
Yes... it look like error.